llama_index/query_engine/knowledge_graph_query_engine.py

```diff
@@ -0,0 +1,178 @@
+        storage_context (Optional[StorageContext]): A storage context to use.
+        refresh_schema (bool): Whether to refresh the schema.
+        verbose (bool): Whether to print intermediate results.
+        **kwargs: Additional keyword arguments.
+
+    """
+
+    def __init__(
+        self,
+        service_context: Optional[ServiceContext] = None,
+        storage_context: Optional[StorageContext] = None,
+        graph_query_synthesis_prompt: Optional[Prompt] = None,
+        graph_response_answer_prompt: Optional[Prompt] = None,
+        refresh_schema: bool = False,
+        verbose: bool = False,
+        **kwargs: Any,
+    ):
+        # Ensure that we have a graph store
+        assert storage_context is not None, "Must provide a storage context."
+        assert (
+            storage_context.graph_store is not None
+        ), "Must provide a graph store in the storage context."
+        self.storage_context = storage_context
+        self.graph_store = storage_context.graph_store
+
+        self.service_context = service_context or ServiceContext.from_defaults()
+
+        # Get Graph Store Type
+        self.graph_store_type = GRAPH_STORE_CLASS_TO_GRAPH_STORE_TYPE[
+            self.graph_store.__class__
+        ]
+
+        # Get Graph schema
+        self.graph_schema = self.graph_store.get_schema(refresh=refresh_schema)
+
+        # Get graph store query synthesis prompt
+        self.graph_query_synthesis_prompt = (
+            graph_query_synthesis_prompt
+            or DEFAULT_NL2GRAPH_PROMPT_MAP[self.graph_store_type]
+        )
+
+        self.graph_response_answer_prompt = (
+            graph_response_answer_prompt or DEFAULT_KG_RESPONSE_ANSWER_PROMPT
+        )
+        self.verbose = verbose
+
+        super().__init__(self.service_context.callback_manager)
+
+    def generate_query(self, query_str: str) -> str:
+        """Generate a Graph Store Query from a query bundle."""
+        # Get the query engine query string
+
+        graph_store_query: str = self.service_context.llm_predictor.predict(
+            self.graph_query_synthesis_prompt,
+            query_str=query_str,
+            schema=self.graph_schema,
+        )
+
+        return graph_store_query
+
+    def _query(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
+        """Query the graph store."""
+        graph_store_query = self.generate_query(query_bundle.query_str)
+        if self.verbose:
+            print_text(f"Graph Store Query: {graph_store_query}\n", color="yellow")
+        logger.info(f"Graph Store Query: {graph_store_query}")
+        # Get the graph store response
+        graph_store_response = self.graph_store.query(query=graph_store_query)
+        if self.verbose:
+            print_text(
+                f"Graph Store Response: {graph_store_response}\n", color="yellow"
+            )
+        logger.info(f"Graph Store Response: {graph_store_response}")
+        response_str = self.service_context.llm_predictor.predict(
```

Comment:
One concern I have here (and it's something that's popped up in the SQL query engine too), is that what if the graph response is huge. Should we be using the `Compact` response synthesizer here instead of a raw predict call, to prevent token overflow?

Reply:
Update, it seems I have to pass string to it, I suspect the citation_query_engine.py is broken as it passed node to it, but eventually it'll call node.strip() which will throw no method strip.

in indices.prompt_help.PromptHelper.repack()

```python
...
combined_str = "\n\n".join([c.strip() for c in text_chunks if c.strip()])
```

Reply:
my bad, I should have specified the method to use

I think the proper method is `CompactAndRefine.get_response()`, which takes a query and a list of text chunks

Reply:
https://github.com/jerryjliu/llama_index/blob/0ce5aa03affbf9943580eef1520ae5b9b588b508/llama_index/response_synthesizers/compact_and_refine.py#L23

Reply:
Sorry, my bad, you were right, synthesize() is different from get_response and now I am using synthesize()